lib/SPIRV/SPIRVReader.cpp

```diff
@@ -2399,6 +2412,55 @@ Value *SPIRVToLLVM::transValueWithoutDecoration(SPIRVValue *BV, Function *F,
   }
 }
 
+CallInst *SPIRVToLLVM::transFixedPointInst(SPIRVInstruction *BI,
+                                           BasicBlock *BB) {
+  // LLVM fixed point functions return value:
+  // iN
+  // Arguments:
+  // A(iN), S(i1), I(i32), rI(i32), Quantization(i32), Overflow(i32)
+
+  // SPIR-V fixed point instruction contains:
+  // <id>ResTy Res<id> <id>InTy In<id> \
+  // Literal S Literal I Literal rI Literal Q Literal O
+
+  Type *RetTy = transType(BI->getType());
+  auto Inst = static_cast<SPIRVFixedPointIntelInst *>(BI);
+
+  IntegerType *Int32Ty = IntegerType::get(*Context, 32);
+  IntegerType *Int1Ty = IntegerType::get(*Context, 1);
+
+  SmallVector<Type *, 7> ArgTys = {transType(Inst->getOperand(1)->getType()),
+                                   Int1Ty,
+                                   Int32Ty,
+                                   Int32Ty,
+                                   Int32Ty,
+                                   Int32Ty};
+
+  FunctionType *FT = FunctionType::get(RetTy, ArgTys, false);
+
+  Op OpCode = Inst->getOpCode();
+  Function *Func = Function::Create(FT, GlobalValue::ExternalLinkage,
```

Comment:
Do you have a test where the same fixed point funciton/opcode is encountered twice? I would expect that `Funciton::Create` will declare the same buit-in each time it is called and it will add `.1, `.2`, etc. suffixes to declaration name. You probably interested in `llvm::Module::getOrInsertFunction`

Reply:
Oh, that's a good point, thanks! I will fix that and add the appropriate test case.

Reply:
This `getOrInsertFunction` is not perfect, though. If the same function is called twice - it works fine, but if the same function has different signatures and return types - a huge bitcast is being created:
`
call i13 bitcast (i5 (i13, i1, i32, i32, i32, i32)* @intel_arbitrary_fixed_sqrt to i13 (i5, i1, i32, i32, i32, i32)*)(i5 %5, i1 false, i32 2, i32 2, i32 0, i32 0)
`
Not sure what is better - bitcast or `.N` at the end of the function name. What do you think of this?

Reply:
It depends on what can be handled by the consumer of that LLVM IR which is translated from the SPIR-V.

I guess that both variants are not really acceptable. Usually, such overloads are distinguished by mangling and `.1`, `.2` is some kind of it, also it is common that LLVM IR representation after the translation is similar to/the same as LLVM IR  representation before translation - probably this is a way you need to go?